## Content
  - [Approach 1: Simple Duplicate Check](#Approach-1-Simple-Duplicate-Check)
  - [Approach 2: Advanced Similarity Check](#Approach-2-Advanced-Similarity-Check)
    - [Threshold = 0.95](#Threshold-0.95)
    - [Threshold = 0.9](#Threshold-0.9)
    - [Threshold = 0.8](#Threshold-0.8)

## Approach 1: Simple Duplicate Check

In [3]:
# 1. Import Libraries
import pandas as pd

# 2. Load Data
df = pd.read_csv('train_split.csv')

# 3. Find duplicates in the 'full_text' column
duplicates = df[df.duplicated(subset=['full_text'], keep=False)]

# 4. Display duplicates
print(f"Number of duplicate entries: {duplicates.shape[0]}")
print(duplicates)


Number of duplicate entries: 0
Empty DataFrame
Columns: [essay_id, full_text, score]
Index: []


## Approach 2: Advanced Similarity Check

In [5]:
## 0.95

In [6]:
# 1. Import Libraries
import pandas as pd
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

# 2. Load Data
df = pd.read_csv('train_split.csv')

# 3. Vectorize the text data using TF-IDF
vectorizer = TfidfVectorizer()
tfidf_matrix = vectorizer.fit_transform(df['full_text'])

# 4. Calculate cosine similarity
cosine_sim = cosine_similarity(tfidf_matrix, tfidf_matrix)

# 5. Find duplicates based on a similarity threshold
threshold = 0.95  # You can adjust this threshold
duplicates = []
seen_indices = set()

for idx, row in enumerate(cosine_sim):
    if idx in seen_indices:
        continue

    similar_indices = [i for i, sim in enumerate(row) if sim > threshold and i != idx]
    if similar_indices:
        duplicates.append((idx, similar_indices))
        seen_indices.update(similar_indices)

# 6. Collect data for display
duplicate_data = []

for dup in duplicates:
    original_idx = dup[0]
    duplicate_indices = dup[1]
    
    original_text = df.iloc[original_idx]['full_text']
    original_score = df.iloc[original_idx]['score']
    original_text_length = len(original_text)
    
    for duplicate_idx in duplicate_indices:
        duplicate_text = df.iloc[duplicate_idx]['full_text']
        duplicate_score = df.iloc[duplicate_idx]['score']
        duplicate_text_length = len(duplicate_text)
        
        duplicate_data.append({
            'Original Index': original_idx,
            'Duplicate Index': duplicate_idx,
            'Original Text': original_text,
            'Duplicate Text': duplicate_text,
            'Original Score': original_score,
            'Duplicate Score': duplicate_score,
            'Original Text Length': original_text_length,
            'Duplicate Text Length': duplicate_text_length
        })

# 7. Create DataFrame from the collected data
df_duplicates = pd.DataFrame(duplicate_data)

# Reorder columns
df_duplicates = df_duplicates[[
    'Original Index', 'Duplicate Index', 
    'Original Text', 'Duplicate Text', 
    'Original Score', 'Duplicate Score', 
    'Original Text Length', 'Duplicate Text Length'
]]

# Count the number of duplicates
num_duplicates = len(duplicate_data)
print(f"Number of duplicate entries: {num_duplicates}")


Number of duplicate entries: 1


In [7]:
# Display the DataFrame
df_duplicates.head(3)

Original Index  Duplicate Index  \
0             789             7538   

                                       Original Text  \
0  Hi, I am PROPER_NAME and I really recomend tha...   

                                      Duplicate Text  Original Score  \
0  Hi, I am luke and I really recomend that you j...               3   

   Duplicate Score  Original Text Length  Duplicate Text Length  
0                3                  1665                   1652

Let's check threshold = 0.9

In [9]:
## 0.9

In [10]:
# 1. Import Libraries
import pandas as pd
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

# 2. Load Data
df = pd.read_csv('train_split.csv')

# 3. Vectorize the text data using TF-IDF
vectorizer = TfidfVectorizer()
tfidf_matrix = vectorizer.fit_transform(df['full_text'])

# 4. Calculate cosine similarity
cosine_sim = cosine_similarity(tfidf_matrix, tfidf_matrix)

# 5. Find duplicates based on a similarity threshold
threshold = 0.9  # You can adjust this threshold
duplicates = []
seen_indices = set()

for idx, row in enumerate(cosine_sim):
    if idx in seen_indices:
        continue

    similar_indices = [i for i, sim in enumerate(row) if sim > threshold and i != idx]
    if similar_indices:
        duplicates.append((idx, similar_indices))
        seen_indices.update(similar_indices)

# 6. Collect data for display
duplicate_data = []

for dup in duplicates:
    original_idx = dup[0]
    duplicate_indices = dup[1]
    
    original_text = df.iloc[original_idx]['full_text']
    original_score = df.iloc[original_idx]['score']
    original_text_length = len(original_text)
    
    for duplicate_idx in duplicate_indices:
        duplicate_text = df.iloc[duplicate_idx]['full_text']
        duplicate_score = df.iloc[duplicate_idx]['score']
        duplicate_text_length = len(duplicate_text)
        
        duplicate_data.append({
            'Original Index': original_idx,
            'Duplicate Index': duplicate_idx,
            'Original Text': original_text,
            'Duplicate Text': duplicate_text,
            'Original Score': original_score,
            'Duplicate Score': duplicate_score,
            'Original Text Length': original_text_length,
            'Duplicate Text Length': duplicate_text_length
        })

# 7. Create DataFrame from the collected data
df_duplicates = pd.DataFrame(duplicate_data)

# Reorder columns
df_duplicates = df_duplicates[[
    'Original Index', 'Duplicate Index', 
    'Original Text', 'Duplicate Text', 
    'Original Score', 'Duplicate Score', 
    'Original Text Length', 'Duplicate Text Length'
]]

# Count the number of duplicates
num_duplicates = len(duplicate_data)
print(f"Number of duplicate entries: {num_duplicates}")


Number of duplicate entries: 2


In [11]:
# Display the DataFrame
df_duplicates.head(4)

Original Index  Duplicate Index  \
0             789             7538   
1            3109            12445   

                                       Original Text  \
0  Hi, I am PROPER_NAME and I really recomend tha...   
1  I really don't like this new computer thing at...   

                                      Duplicate Text  Original Score  \
0  Hi, I am luke and I really recomend that you j...               3   
1  I really don't like this new computer thing at...               3   

   Duplicate Score  Original Text Length  Duplicate Text Length  
0                3                  1665                   1652  
1                3                  1311                   1297

Thoughts
- this option seems the best for removal. Scores are similar, length too.
- which one to choose for removal? Texts were analyzed additionally, and those which contain [proper name] inside will be removed. These are 789, 3109

In [13]:
## 0.8

In [14]:
# 1. Import Libraries
import pandas as pd
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

# 2. Load Data
df = pd.read_csv('train_split.csv')

# 3. Vectorize the text data using TF-IDF
vectorizer = TfidfVectorizer()
tfidf_matrix = vectorizer.fit_transform(df['full_text'])

# 4. Calculate cosine similarity
cosine_sim = cosine_similarity(tfidf_matrix, tfidf_matrix)

# 5. Find duplicates based on a similarity threshold
threshold = 0.80  # You can adjust this threshold
duplicates = []
seen_indices = set()

for idx, row in enumerate(cosine_sim):
    if idx in seen_indices:
        continue

    similar_indices = [i for i, sim in enumerate(row) if sim > threshold and i != idx]
    if similar_indices:
        duplicates.append((idx, similar_indices))
        seen_indices.update(similar_indices)

# 6. Collect data for display
duplicate_data = []

for dup in duplicates:
    original_idx = dup[0]
    duplicate_indices = dup[1]
    
    original_text = df.iloc[original_idx]['full_text']
    original_score = df.iloc[original_idx]['score']
    original_text_length = len(original_text)
    
    for duplicate_idx in duplicate_indices:
        duplicate_text = df.iloc[duplicate_idx]['full_text']
        duplicate_score = df.iloc[duplicate_idx]['score']
        duplicate_text_length = len(duplicate_text)
        
        duplicate_data.append({
            'Original Index': original_idx,
            'Duplicate Index': duplicate_idx,
            'Original Text': original_text,
            'Duplicate Text': duplicate_text,
            'Original Score': original_score,
            'Duplicate Score': duplicate_score,
            'Original Text Length': original_text_length,
            'Duplicate Text Length': duplicate_text_length
        })

# 7. Create DataFrame from the collected data
df_duplicates = pd.DataFrame(duplicate_data)

# Reorder columns
df_duplicates = df_duplicates[[
    'Original Index', 'Duplicate Index', 
    'Original Text', 'Duplicate Text', 
    'Original Score', 'Duplicate Score', 
    'Original Text Length', 'Duplicate Text Length'
]]

# Count the number of duplicates
num_duplicates = len(duplicate_data)
print(f"Number of duplicate entries: {num_duplicates}")


Number of duplicate entries: 7


In [15]:
# Display the DataFrame
df_duplicates.head(16)

Original Index  Duplicate Index  \
0             789             7538   
1             928             4511   
2             928             8766   
3            3109            12445   
4            3385             5672   
5            7325             9671   
6           11339            13197   

                                       Original Text  \
0  Hi, I am PROPER_NAME and I really recomend tha...   
1  I think the Electoral college is a good way to...   
2  I think the Electoral college is a good way to...   
3  I really don't like this new computer thing at...   
4  A new hom\n\nwhould you send someone to explor...   
5  teacher asks:\n\nWhat is an electoral college?...   
6  Dear senator,\n\nWhat are the chances in favor...   

                                      Duplicate Text  Original Score  \
0  Hi, I am luke and I really recomend that you j...               3   
1  Dear State Senator jan,28 2015\n\nI'm not in f...               1   
2  I think that we should keep the Electoral Coll...               1   
3  I really don't like this new computer thing at...               3   
4  Benefits of Researching a New planet\n\nWhould...               2   
5  In the first source its explain what a Elector...               1   
6  Dear state senator i am writing to you because...               2   

   Duplicate Score  Original Text Length  Duplicate Text Length  
0                3                  1665                   1652  
1                2                  3561                   3636  
2                1                  3561                   2978  
3                3                  1311                   1297  
4                4                  1750                   2252  
5                1                  1777                   2576  
6                2                  1417                   1481

Texts have different scores, noticeably due to different lengths; hence this option will be skipped. It appears that students might have copied essays from friends and enriched them.